### PR TITLE
[Core/Scripts]: Add locale support to Autobroadcast

### PR DESCRIPTION
**Pull request by lineagedr**
## Changes Proposed:
- Move autobroadcast to it's own file.
- Add support for announcing text for different client locales.
- Move WorldTextBuilder class definition to World.h Nothing really changes except we can now access it outside of World.cpp But we still need to do it through the Acore namespace.

## Issues Addressed:

- Closes Autobroadcast part of https://github.com/azerothcore/azerothcore-wotlk/issues/2856

## Tests Performed:

- Tested with en_US client and the rest of locales were viewed in the console and they seemed to display.

## How to Test the Changes:

1. Add your custom locale message in the autobroadcast table according to your client's locale.
2. Login and see if you get the message.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Autobroadcast/AutobroadcastMgr.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+class AC_GAME_API AutobroadcastMgr
+{
+public:
+    static AutobroadcastMgr* instance();
+
+    void Load();
+    void Send();
+
+private:
+    typedef std::map<LocaleConstant, std::string> AutobroadcastTextLocale;
```
Comment (Winfidonarleyan) [inline src/server/game/Autobroadcast/AutobroadcastMgr.h:26]: Why used typedef?

Comment (lineagedr) [reply src/server/game/Autobroadcast/AutobroadcastMgr.h:26]: Readability.

File: data/sql/updates/pending_db_auth/rev_1629280349876853000.sql [hunk 2]
```diff
@@ -0,0 +1,13 @@
+INSERT INTO `version_db_auth` (`sql_rev`) VALUES ('1629280349876853000');
+
+ALTER TABLE autobroadcast CHANGE COLUMN `text` `text_enUS` LONGTEXT NOT NULL;
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_auth/rev_1629280349876853000.sql:3]: I'd prefer to keep this as "text" or "text_default" as imo it should default to this if client locale isn't found or something.

Mainly thinking of acore_string that has "content_defualt"

File: src/server/game/Autobroadcast/AutobroadcastMgr.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ * Copyright (C) 2016+ AzerothCore <www.azerothcore.org>, released under GNU AGPL v3 license: https://github.com/azerothcore/azerothcore-wotlk/blob/master/LICENSE-AGPL3
+ */
+
+#ifndef _AUTOBROADCAST_MANAGER_H_
+#define _AUTOBROADCAST_MANAGER_H_
+
+#include "Common.h"
+#include <map>
+
+enum class AnnounceType : uint8
```
Comment (Kitzunu) [inline src/server/game/Autobroadcast/AutobroadcastMgr.h:11]: nice

File: src/server/game/Autobroadcast/AutobroadcastMgr.cpp [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    };
+
+    for (const auto& [localeId, message] : msg)
+    {
+        if (message.empty())
+        {
+            continue;
+        }
+
+        switch (announceType)
```
Comment (Kitzunu) [inline src/server/game/Autobroadcast/AutobroadcastMgr.cpp:147]:
```suggestion
        switch (announceType) //@todo Change to using enum with c++20
```

Comment (Kitzunu) [reply src/server/game/Autobroadcast/AutobroadcastMgr.cpp:147]: For me to remember for my PR :D

File: data/sql/updates/pending_db_auth/rev_1629280349876853000.sql (hunk 2 above)
Comment (lineagedr) [reply data/sql/updates/pending_db_auth/rev_1629280349876853000.sql:3]: The client locale will always be found. The only usability of a text_default that I can think of would be if the person is too lazy to fill the rest of the text locale columns.

**Comment by Maurowin** [comment]
Is this PR only for autobroadcasts or does it also take into account MOTD?

**Comment by lineagedr** [comment]
@Maurowin Only annouce.

File: data/sql/updates/pending_db_auth/rev_1629280349876853000.sql (hunk 2 above)
Comment (Nefertumm) [reply data/sql/updates/pending_db_auth/rev_1629280349876853000.sql:3]:
```suggestion
ALTER TABLE `autobroadcast` CHANGE COLUMN `text` `text_enUS` LONGTEXT NOT NULL;
```

File: data/sql/updates/pending_db_auth/rev_1629280349876853000.sql [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+ALTER TABLE autobroadcast
```
Comment (Nefertumm) [inline data/sql/updates/pending_db_auth/rev_1629280349876853000.sql:5]:
```suggestion
ALTER TABLE `autobroadcast`
```

File: src/server/game/Autobroadcast/AutobroadcastMgr.cpp [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+    const auto sendAsNotification = [](LocaleConstant locale, const std::string_view message)
+    {
+        WorldPacket data(SMSG_NOTIFICATION, (message.size() + 1));
+        data << message;
+
+        for (const auto& [id, session] : sWorld->GetAllSessions())
+        {
+            if (!session || !session->GetPlayer() || !session->GetPlayer()->IsInWorld() || session->GetSessionDbLocaleIndex() != locale)
+                continue;
```
Comment (Nefertumm) [inline src/server/game/Autobroadcast/AutobroadcastMgr.cpp:116]: Brackets here

File: src/server/game/Autobroadcast/AutobroadcastMgr.cpp [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    {
+        va_list ap;
+        va_start(ap, string_id);
+
+        Acore::WorldTextBuilder wt_builder(string_id, &ap);
+        Acore::LocalizedPacketListDo<Acore::WorldTextBuilder> wt_do(wt_builder);
+        for(const auto& [id, session] : sWorld->GetAllSessions())
+        {
+            if (!session || !session->GetPlayer() || !session->GetPlayer()->IsInWorld() || session->GetSessionDbLocaleIndex() != locale)
+                continue;
```
Comment (Nefertumm) [inline src/server/game/Autobroadcast/AutobroadcastMgr.cpp:132]: Brackets

File: src/server/game/Autobroadcast/AutobroadcastMgr.cpp [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    for (const auto& [localeId, message] : msg)
+    {
+        if (message.empty())
+        {
+            continue;
+        }
+
+        switch (announceType) //@todo Change to using enum with c++20
+        {
+        case AnnounceType::WORLD:
```
Comment (Nefertumm) [inline src/server/game/Autobroadcast/AutobroadcastMgr.cpp:149]: Fix indent here

File: src/server/game/World/World.h [hunk 9]
```diff
@@ -148,6 +148,24 @@ struct PetitionData
 {
 };
 
+namespace Acore
+{
+    class WorldTextBuilder
+    {
+    public:
```
Comment (Nefertumm) [inline src/server/game/World/World.h:155]: Tabs here too maybe?

**Review by Nefertumm — APPROVED** [review]
Looking good, just change what I marked to match standards 👍

**Comment by FrancescoBorzi** [comment]
@lineagedr what's the status of this PR?

**Comment by Winfidonarleyan** [comment]
> @lineagedr what's the status of this PR?

Can close

**Pull request closed without merging**